### PR TITLE
Read `game.properties` and `ext.properties` as metadata for `game.project`

### PR DESCRIPTION
**Pull request by AGulev**
From now on it's possible to create metadata for the `game.project` file in the project itself and in an extension.

The project file should have the name `game.properties` in the root folder (near `game.project` file) for an extension it's `ext.properties` next to `ext.manifest`.

This file has `ini` format where attributes may be specified after dots:
```
[my_category]
my_property.private = 1
...
```
The default meta file that is always applied and available [here](https://github.com/defold/defold/blob/dev/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties)

Currently available are the following attributes:
```
// `type` - used for the value string parsing (only in bob.jar for now)
my_property.type = string // one of the follwoing values: bool, string, number, integer, string_array, resource

// `help` - used as help tip in the editor (not used for now)
my_property.help = string

// `default` - value used as default if user didn't set value manually (only in bob.jar for now)
my_property.default = string

// `private` - private value used during the bundle process but will be removed from the bundle itself
my_property.private = 1 // boolean value 1 or 0

``` 

At the moment meta properties are used only in `bob.jar` when bundling an app, but later will be parsed by the editor and represented in the `game.project` viewer.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java [hunk 1]
```diff
@@ -333,15 +526,37 @@ static private Map<String, Map<String, String>> doLoad(InputStream in, KeyValueF
...
+                            String[] keyAndAttr = fullKey.split("\\.");
+                            String key = keyAndAttr[0];
+                            ProjectProperty prop = propGroup.get(key);
+                            if (prop == null) {
+                                prop = new ProjectProperty();
+                            }
+                            prop.setAttribute(keyAndAttr[1], value);
+                            propGroup.put(key, prop);
+                        }
+                        else if (!isMeta && fullKey.contains("#")) {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java:541]: universal parsing of array values (with index after `#`)

**Review by britzl — DISMISSED** [review]
It's a pretty big change. It looks good though.

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/bundle/test/BundleHelperTest.java [hunk 2]
```diff
@@ -54,23 +53,6 @@ public void tearDown() throws Exception {
         this.mp.unmount();
     }
 
-    void testValue(String p, String c, String k, Object expected) throws IOException, ParseException {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/bundle/test/BundleHelperTest.java:57]: Now it's a part of BobProjectPropertiesTest.java

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java [hunk 3]
```diff
@@ -108,6 +108,7 @@ public static void throwIfCanceled(ICanceled canceled) {
 
     public BundleHelper(Project project, Platform platform, File bundleDir, String variant) throws CompileExceptionError {
         this.projectProperties = project.getProjectProperties();
+        this.propertiesMap = this.projectProperties.createTypedMap(new BobProjectProperties.PropertyType[]{BobProjectProperties.PropertyType.BOOL});
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java:111]: We can specify which types should be parsed for the map we wanna get

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java [hunk 4]
```diff
@@ -148,75 +143,15 @@ public static String projectNameToBinaryName(String projectName) {
...
-            if (value != null) {
-                return value.equals("1");
-            } else {
-                return false;
-            }
-        }
-        return value;
-    }
-
-    public static Map<String, Map<String, Object>> createPropertiesMap(BobProjectProperties projectProperties) throws IOException {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java:162]: now it's part of `BobProjectProperties.java`

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java [hunk 5]
```diff
@@ -148,75 +143,15 @@ public static String projectNameToBinaryName(String projectName) {
...
+    private IResource getResource(String category, String key) throws IOException {
+        String val = this.projectProperties.getStringValue(category, key);
+        if (val != null && val.trim().length() > 0) {
+            return project.getResource(val);
         }
         throw new IOException(String.format("No resource found for %s.%s", category, key));
     }
 
-    public String formatResource(Map<String, Object> properties, IResource resource) throws IOException {
+    private String formatResource(Map<String, Object> properties, IResource resource) throws IOException {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/BundleHelper.java:154]: more private function (easier to read code )

**Review by britzl — DISMISSED** [review]

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/test/ProjectBuildTest.java [hunk 6]
```diff
@@ -189,6 +189,19 @@ public void testGameProperties() throws IOException, ConfigurationException, Com
...
+        projectName = "String Array";
+        createDefaultFiles();
+        createFile(contentRoot, "game.project", "[project]\ntitle = " + projectName +
+            "\ncustom_string_list#0 = http://test.com/test.zip\ncustom_string_list#2 = http://test.com/test2.zip\ncustom_string_list#1 = http://test.com/test1.zip\n");
+        build();
+        BobProjectProperties outputProps = new BobProjectProperties();
+        outputProps.load(new FileInputStream(new File(contentRoot + "/build/game.projectc")));
+
+        checkProjectSettingArray(outputProps, "project", "custom_string_list", new String[]{"http://test.com/test.zip", "http://test.com/test1.zip", "http://test.com/test2.zip"});
+    }
```
[the remark this answers is not in the document]
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/test/ProjectBuildTest.java:203]: I agree, done! Thx

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java [hunk 7]
```diff
@@ -25,34 +25,299 @@
...
+        public void overrideBy(ProjectProperty prop) {
+            try {
+                Field[] fields = ProjectProperty.class.getDeclaredFields();
+                for(Field f : fields) {
+                    Class t = f.getType();
+                    Object v = f.get(prop);
+                    if ((v != null) && (t.isEnum() || t.equals(String.class) || t.equals(Boolean.class))) {
+                        f.set(this, v);
+                        if (f.getName() == "value") {
+                            // it's important to reset old parsed values from value Array in value overwritten
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java:152]: "in value" -> "if value is" ?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 8]
```diff
@@ -534,36 +534,57 @@ public void clearProjectProperties() {
...
-            Project.loadPropertyFile(properties, resource.getAbsPath());
+            // load meta.properties embeded in bob.jar
+            properties.loadDefaultMetaFile();
+            // load property files from extensions
+            List<String> extensionFolders = ExtenderUtil.getExtensionFolders(project);
+            if (!extensionFolders.isEmpty()) {
+                for (String extension : extensionFolders) {
+                    IResource resource = project.getResource(extension + "/" + BobProjectProperties.PROPERTIES_EXTENSION_FILE);
+                    if (resource.exists()) {
+                        // resources from extensions in ZIP files can't be read as files, but getContent() wroks fine
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:570]: "wrok" -> "work"

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java [hunk 9]
```diff
@@ -25,34 +25,299 @@
...
+        load(in, false);
+    }
+
+    /**
+     * Load default properties from the internal meta.properties file
+     * @throws IOException
+     * @throws ParseException
+     */
+    public void loadDefaultMetaFile()  throws IOException, ParseException {
+        InputStream is = Bob.class.getResourceAsStream("meta.properties");
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java:279]: Should this name also have a variable?
E:g. like `PROPERTIES_EXTENSION_FILE` does.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java (hunk 1 above)
Comment (JCash) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java:566]: I'm not sure about this whole if-statement.
If the key contains a `.`, then we'll never get here?

**Review by JCash — CHANGES_REQUESTED** [review]

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java (hunk 1 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java:566]: yes, you right
you can't specify attribute for an element in array
so, if it's attribute (it has `.` in it) then we parse it as an attribute

**Pull request merged**